File: client/cli/src/commands/generate.ts
import type { Argv, CommandBuilder, CommandModule } from 'yargs';
import { loadBudget } from '@tally/lib/data/loader';
import { Account } from '@tally/lib/core/account';
import { Balance, Type as BalanceType } from '@tally/lib/core/balance';
import { Budget } from '@tally/lib/core/budget';
import { Month } from '@tally/lib/core/month';
import { Transfer } from '@tally/lib/core/transfer';

type GenerateOptions = {
  account: string;
  startMonth: Month;
  useTransfers: boolean;
  showTransfers: boolean;
};

const command: string = 'generate <account>';
const desc: string = 'Generate balances record based on transfers.';

const builder: CommandBuilder<unknown, GenerateOptions> = (yargs: Argv<unknown>) =>
  yargs
    .options({
      'start-month': { type: 'string', demandOption: true },
      'use-transfers': {
        type: 'boolean',
        default: false,
        description: 'Ignore balance entries and use existing transfer to compute them.',
      },
      'show-transfers': {
        type: 'boolean',
        default: false,
        description: 'Show existing transfers for debugging.',
      },
      'with-annual-flush': {
        type: 'boolean',
        default: false,
        description: 'Generate annual flush transfers. (Make sure to remove existing transfers)',
      },
    })
    .positional('account', { type: 'string', demandOption: true })
    .coerce('start-month', Month.fromString) as unknown as Argv<GenerateOptions>;

export const generate: CommandModule<unknown, GenerateOptions> = {
  command,
  describe: desc,
  builder,
  handler: async ({
    account,
    startMonth,
    useTransfers,
    showTransfers,
    withAnnualFlush,
  }): Promise<void> => {
    const budget: Budget = (await loadBudget()).budget;
    process.stdout.write(
      `Generating balances for ${account} starting from ${startMonth?.toString()}!\n`
    );
    const acct: Account | undefined = budget.accounts.get(account);
    if (!acct) {
      process.stderr.write(`The account "${account}" does not exist.\n`);
      process.exit(1);
    }
    const accountBalances: Map<string, Balance> = budget.balances.get(account) ?? new Map();
    const accountTransfers: Map<string, Set<Transfer>> = budget.transfers.get(account) ?? new Map();

    // Find minimum/maximum month used by this account.
    const accountMonths = [...accountBalances.keys(), ...accountTransfers.keys()].map((val) =>
      Month.fromString(val)
    );
    const minMonth = accountMonths.reduce((prev, current) =>
      prev.compareTo(current) < 0 ? prev : current
    );
    const maxMonth = accountMonths.reduce((prev, current) =>
      prev.compareTo(current) > 0 ? prev : current
    );
    if (startMonth.compareTo(minMonth) < 0) {
      startMonth = minMonth.previous();
    }
    if (startMonth.compareTo(maxMonth) > 0) {
      process.stderr.write(
        `The account "${account}" has no balances or transactions after ${maxMonth}.\n`
      );
      process.exit(1);
    }

    // Compute max padding for amounts to be right aligned.
    const ammountLengths: number[] = budget.months.map(
      (m) => `${budget.balances.get(account)?.get(m.toString())?.amount ?? 0}`.length
    );
    const padAmtLength = Math.max(...ammountLengths) + 2; // 1 extra leading space plus "."

    const lines: string[] = [];
    const flushLines: string[] = [];
    // For each month, compute running predicted balance and actual balance.
    let predictedBalance: Balance | undefined = undefined;
    for (
      let currentMonth = startMonth;
      currentMonth.compareTo(maxMonth) <= 0;
      currentMonth = currentMonth.next()
    ) {
      const recordedBalance = accountBalances.get(currentMonth.toString());
      let currentBalance: Balance | undefined = useTransfers
        ? predictedBalance ?? recordedBalance
        : recordedBalance ?? predictedBalance;
      if (!currentBalance) {
        predictedBalance = undefined;
        const prevTransfers = accountTransfers.get(currentMonth.previous().toString()) ?? new Set();
        lines.push(
          `  - { grp: ${currentMonth} } # has no balance and had ${prevTransfers.size} transfers.`
        );
        continue;
      }
      lines.push(printBalanceLine(currentMonth, currentBalance, padAmtLength));
      // If there is disaggreement print it as a comment.
      if (predictedBalance && predictedBalance.amount !== currentBalance.amount) {
        const predictedAmtValue = (predictedBalance.amount / 100).toFixed(2).padStart(padAmtLength);
        const diffAmtValue = ((currentBalance.amount - predictedBalance.amount) / 100)
          .toFixed(2)
          .padStart(padAmtLength);
        lines[lines.length - 1] += ` # predicted ${predictedAmtValue} unaccounted ${diffAmtValue}`;
      }
      const transfers: Set<Transfer> = accountTransfers.get(currentMonth.toString()) ?? new Set();
      if (showTransfers) {
        for (const transfer of transfers) {
          lines.push(
            `    ${transfer.toMonth} ${transfer.fromAccount.name} --> ${transfer.toAccount.name} ${transfer.balance}`
          );
        }
      }
      // Find first date of next month transcation, our predicated balance cannot be older.
      const nextTransfers = accountTransfers.get(currentMonth.next().toString()) ?? new Set();
      const minDateNextMonthTransfer: Date | undefined = nextTransfers.size
        ? [...nextTransfers]
            .map((t) => t.balance.date)
            .reduce((prev, curr) => (prev.getTime() < curr.getTime() ? prev : curr))
        : undefined;
      if (withAnnualFlush && currentMonth.month === 0) {
        const firstMonth = currentMonth;
        const amtValue = (currentBalance.amount / 100).toFixed(2).padStart(padAmtLength);
        flushLines.push(
          `    - { grp: ${firstMonth}, date: ${
            firstMonth.year
          }-01-02, camt: ${amtValue}, desc: "Total for ${currentMonth.year - 1}" }`
        );
        currentBalance = new Balance(0, currentBalance.date, currentBalance.type);
      }
      predictedBalance = nextBalance(
        account,
        currentBalance,
        transfers,
        minDateNextMonthTransfer,
        useTransfers ? BalanceType.CONFIRMED : BalanceType.PROJECTED
      );
    }

    lines.reverse();
    for (const line of lines) {
      process.stdout.write(line + '\n');
    }
    flushLines.reverse();
    for (const line of flushLines) {
      process.stdout.write(line + '\n');
    }
    process.exit(0);
  },
};

function printBalanceLine(month: Month, balance: Balance, padAmtLength: number) {
  const amtPrefix = balance.type === BalanceType.PROJECTED ? 'pamt' : 'camt';
  const amtValue = (balance.amount / 100).toFixed(2).padStart(padAmtLength);
  const dateValue = balance.date.toISOString().slice(0, 10);
  return `  - { grp: ${month}, date: ${dateValue}, ${amtPrefix}: ${amtValue} }`;
}

function nextBalance(
  accountName: string,
  startBalance: Balance,
  transfers: Set<Transfer>,
  minDateNextMonthTransfer: Date | undefined,
  balanceType: BalanceType
): Balance | undefined {
  //if (transfers.size === 0) { return undefined; }
  // The balance date is set to max transfer date,
  // but make sure it is not lower then next month start by default.
  const nextDate = new Date(
    Date.UTC(
      startBalance.date.getUTCFullYear(),
      startBalance.date.getUTCMonth() + 1,
      startBalance.date.getUTCDate()
    )
  );
  let balance = new Balance(startBalance.amount, nextDate, balanceType);
  for (const transfer of transfers) {
    if (transfer.toAccount.name === accountName) {
      balance = Balance.add(balance, transfer.balance);
    } else {
      balance = Balance.subtract(balance, transfer.balance);
    }
  }
  // Make sure next balance start does not exceed its minimum transfer date.
  if (minDateNextMonthTransfer && balance.date.getTime() > minDateNextMonthTransfer.getTime()) {
    return new Balance(balance.amount, minDateNextMonthTransfer, balance.type);
  }
  return balance;
}

File: server/kt/src/main/kotlin/com/cherba29/tally/cli/cmds/Generate.kt
package com.cherba29.tally.cli.cmds

import com.cherba29.tally.core.Balance
import com.cherba29.tally.core.Month
import com.cherba29.tally.core.Transfer
import com.cherba29.tally.data.Loader
import com.cherba29.tally.data.watchedEventFlow
import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.Context
import com.github.ajalt.clikt.core.UsageError
import com.github.ajalt.clikt.parameters.arguments.argument
import com.github.ajalt.clikt.parameters.options.convert
import com.github.ajalt.clikt.parameters.options.flag
import com.github.ajalt.clikt.parameters.options.option
import com.github.ajalt.clikt.parameters.options.required
import com.github.ajalt.clikt.parameters.types.path
import kotlinx.coroutines.runBlocking
import kotlinx.datetime.DatePeriod
import kotlinx.datetime.LocalDate
import kotlinx.datetime.plus
import kotlin.io.path.extension
import kotlin.io.path.pathString

class Generate : CliktCommand() {
  override fun help(context: Context) = "Full report"

  val tallyPath by option(envvar = "TALLY_PATH").path(mustExist = true).required()

  val account by argument(help = "Account name")
  val startMonth by option(
    "--start-month",
    help = "Inclusive start month, eg Apr2026"
  ).convert { Month.fromString(it) }.required()

  val useTransfers by option(
    "--use-transfers",
    help = "Ignore balance entries and use existing transfer to compute them."
  ).flag(default = false)

  val showTransfers by option(
    "--show-transfers",
    help = "Show existing transfers for debugging."
  ).flag(default = false)

  val withAnnualFlush by option(
    "--with-annual-flush",
    help = "Generate annual flush transfers. (Make sure to remove existing transfers)."
  ).flag(default = false)

  override fun run() {
    echo("Generating balances for $account starting from $startMonth for $tallyPath")
    val loader = Loader(tallyPath.watchedEventFlow {
      it.extension == "yaml" && !ignorePathRegex.containsMatchIn(it.pathString)
    })
    val budget = runBlocking { loader.budget() }.budget
    val acct = budget.accounts.keys.find { it.name == account } ?: throw UsageError("Account $account not found")
    val accountBalances: Map<Month, Balance> = budget.balances[acct] ?: mapOf()
    val accountTransfers: Map<Month, List<Transfer>> = budget.transfers[acct] ?: mapOf()

    // Find minimum/maximum month used by this account.
    val accountMonths = accountBalances.keys + accountTransfers.keys
    val minMonth = accountMonths.min()
    val maxMonth = accountMonths.max()
    val firstMonth = if (startMonth < minMonth) minMonth.previous() else startMonth
    if (firstMonth > maxMonth) {
      throw UsageError("The account $account has no balances or transactions after $maxMonth.")
    }
    // Compute max padding for amounts to be right aligned.
    val amountLengths: List<Int> = budget.months.toList().map {
      (budget.balances[acct]?.get(it)?.amount ?: 0).toString().length
    }
    val padAmtLength = amountLengths.max() + 2  // 1 extra leading space plus "."

    val lines = mutableListOf<String>()
    val flushLines = mutableListOf<String>()

    // For each month, compute running predicted balance and actual balance.
    var predictedBalance: Balance? = null
    for (currentMonth in firstMonth..maxMonth) {
      val recordedBalance = accountBalances[currentMonth]
      var currentBalance: Balance? =
        if (useTransfers) (predictedBalance ?: recordedBalance) else (recordedBalance ?: predictedBalance)
      if (currentBalance == null) {
        predictedBalance = null
        val prevTransfers = accountTransfers[currentMonth.previous()] ?: setOf()
        lines.add(
          "  - { grp: $currentMonth } # has no balance and had ${prevTransfers.size} transfers."
        )
        continue
      }
      lines.add(printBalanceLine(currentMonth, currentBalance, padAmtLength))
      // If there is disagreement print it as a comment.
      if (predictedBalance != null && predictedBalance.amount != currentBalance.amount) {
        val predictedAmtValue = predictedBalance.amount.asAmount().padStart(padAmtLength)
        val diffAmtValue = (currentBalance.amount - predictedBalance.amount).asAmount().padStart(padAmtLength)
        lines[lines.size - 1] += " # predicted $predictedAmtValue unaccounted $diffAmtValue"
      }
      val transfers: Set<Transfer> = accountTransfers[currentMonth]?.toSet() ?: setOf()
      if (showTransfers) {
        for (transfer in transfers) {
          lines.add(
            "    ${transfer.toMonth} ${transfer.fromAccount.nodeId.name} --> ${transfer.toAccount.nodeId.name} ${transfer.balance}"
          )
        }
      }
      // Find first date of next month transaction, our predicated balance cannot be older.
      val nextTransfers = accountTransfers[currentMonth.next()] ?: listOf()
      val minDateNextMonthTransfer: LocalDate? = if (nextTransfers.isNotEmpty()) {
        nextTransfers.maxBy { it.balance.date }.balance.date
      } else null
      if (withAnnualFlush && currentMonth.month == 0) {
        val amtValue = currentBalance.amount.asAmount().padStart(padAmtLength)
        flushLines.add(
          "    - { grp: $currentMonth, date: ${currentMonth.year}-01-02, camt: ${amtValue}, " +
              "desc: \"Total for ${currentMonth.year - 1}\" }"
        )
        currentBalance = Balance(0, currentBalance.date, currentBalance.type)
      }
      predictedBalance = nextBalance(
        account,
        currentBalance,
        transfers,
        minDateNextMonthTransfer,
        if (useTransfers) Balance.Type.CONFIRMED else Balance.Type.PROJECTED
      )
    }

    lines.reverse()
    for (line in lines) {
      echo(line)
    }
    flushLines.reverse()
    for (line in flushLines) {
      echo(line)
    }
  }

  companion object {
    private val ignorePathRegex = Regex("(^_)|(/_)")
    private fun Int.asAmount(): String = "%.2f".format(this / 100.0)

    private fun printBalanceLine(month: Month, balance: Balance, padAmtLength: Int): String {
      val amtPrefix = if (balance.type == Balance.Type.PROJECTED) "pamt" else "camt"
      val amtValue = balance.amount.asAmount().padStart(padAmtLength)
      val dateValue = balance.date.toString()
      return "  - { grp: ${month}, date: ${dateValue}, ${amtPrefix}: $amtValue }"
    }

    private fun nextBalance(
      accountName: String,
      startBalance: Balance,
      transfers: Set<Transfer>,
      minDateNextMonthTransfer: LocalDate?,
      balanceType: Balance.Type
    ): Balance? {
      //if (transfers.size === 0) { return undefined; }
      // The balance date is set to max transfer date,
      // but make sure it is not lower than next month start by default.
      val nextDate = startBalance.date + DatePeriod(months = 1)
      var balance: Balance? = Balance(startBalance.amount, nextDate, balanceType)
      for (transfer in transfers) {
        balance = if (transfer.toAccount.nodeId.name == accountName) {
          Balance.add(balance, transfer.balance)
        } else {
          Balance.subtract(balance, transfer.balance)
        }
      }
      // Make sure next balance start does not exceed its minimum transfer date.
      if (minDateNextMonthTransfer != null && balance!!.date > minDateNextMonthTransfer) {
        return Balance(balance.amount, minDateNextMonthTransfer, balance.type)
      }
      return balance
    }
  }
}